src/slice.rs

```diff
@@ -240,220 +260,357 @@ impl From<RangeFull> for Slice {
+    type OutDim: Dimension;
+
+    fn in_ndim(&self) -> usize;
+
+    fn out_ndim(&self) -> usize;
+}
+
+macro_rules! impl_canslice_samedim {
+    ($in_dim:ty) => {
+        unsafe impl<T, Dout> CanSlice<$in_dim> for SliceInfo<T, $in_dim, Dout>
+        where
+            T: AsRef<[AxisSliceInfo]>,
+            Dout: Dimension,
+        {
+            type OutDim = Dout;
+
+            fn in_ndim(&self) -> usize {
+                self.in_ndim()
+            }
+
+            fn out_ndim(&self) -> usize {
+                self.out_ndim()
             }
         }
     };
 }
-impl_sliceorindex_from_index!(isize);
-impl_sliceorindex_from_index!(usize);
-impl_sliceorindex_from_index!(i32);
+impl_canslice_samedim!(Ix0);
+impl_canslice_samedim!(Ix1);
+impl_canslice_samedim!(Ix2);
+impl_canslice_samedim!(Ix3);
+impl_canslice_samedim!(Ix4);
+impl_canslice_samedim!(Ix5);
+impl_canslice_samedim!(Ix6);
+
+unsafe impl<T, Din, Dout> CanSlice<IxDyn> for SliceInfo<T, Din, Dout>
+where
+    T: AsRef<[AxisSliceInfo]>,
+    Din: Dimension,
+    Dout: Dimension,
+{
+    type OutDim = Dout;
+
+    fn in_ndim(&self) -> usize {
+        self.in_ndim()
+    }
+
+    fn out_ndim(&self) -> usize {
+        self.out_ndim()
+    }
+}
+
+unsafe impl CanSlice<IxDyn> for [AxisSliceInfo] {
+    type OutDim = IxDyn;
+
+    fn in_ndim(&self) -> usize {
+        self.iter().filter(|s| !s.is_new_axis()).count()
+    }
+
+    fn out_ndim(&self) -> usize {
+        self.iter().filter(|s| !s.is_index()).count()
+    }
+}
 
 /// Represents all of the necessary information to perform a slice.
 ///
-/// The type `T` is typically `[SliceOrIndex; n]`, `[SliceOrIndex]`, or
-/// `Vec<SliceOrIndex>`. The type `D` is the output dimension after calling
-/// [`.slice()`].
+/// The type `T` is typically `[AxisSliceInfo; n]`, `[AxisSliceInfo]`, or
+/// `Vec<AxisSliceInfo>`. The type `Din` is the dimension of the array to be
+/// sliced, and `Dout` is the output dimension after calling [`.slice()`]. Note
+/// that if `Din` is a fixed dimension type (`Ix0`, `Ix1`, `Ix2`, etc.), the
+/// `SliceInfo` instance can still be used to slice an array with dimension
+/// `IxDyn` as long as the number of axes matches.
 ///
 /// [`.slice()`]: struct.ArrayBase.html#method.slice
 #[derive(Debug)]
-#[repr(C)]
-pub struct SliceInfo<T: ?Sized, D: Dimension> {
-    out_dim: PhantomData<D>,
+#[repr(transparent)]
+pub struct SliceInfo<T: ?Sized, Din: Dimension, Dout: Dimension> {
+    in_dim: PhantomData<Din>,
+    out_dim: PhantomData<Dout>,
     indices: T,
 }
 
-impl<T: ?Sized, D> Deref for SliceInfo<T, D>
+impl<T: ?Sized, Din, Dout> Deref for SliceInfo<T, Din, Dout>
 where
-    D: Dimension,
+    Din: Dimension,
+    Dout: Dimension,
 {
     type Target = T;
     fn deref(&self) -> &Self::Target {
         &self.indices
     }
 }
 
-impl<T, D> SliceInfo<T, D>
+impl<T, Din, Dout> SliceInfo<T, Din, Dout>
 where
-    D: Dimension,
+    T: AsRef<[AxisSliceInfo]>,
```

Comment:
This raises my `AsRef` concern - an unfaithful/evil impl of `AsRef` can return different slices at different times, we can't trust `AsRef` for `unsafe` code blocks - I think it means the check we do in `SliceInfo::new` is not effectively guarding the invariant we need because `.as_ref()` can later return a different slice. 🙁  This is one of my pet peeves - can't trust `AsRef` for unsafe code, unfortunately.

Reply:
Oh, good point. I've changed `SliceInfo::new` to be `unsafe` and added `TryFrom` implementations for a whitelisted set of types (slices, `Vec`, and fixed-length arrays) which we should be able to depend on having a correct `AsRef` impl.

It would be nice to make `CanSlice` and the `SliceInfo` constructors safe, i.e. move the responsibility for the necessary checks to the slicing methods. Modifying the `.slice*()` methods would be fairly simple. (The primary change would be to make sure that they call `.as_ref()` only once.) However, modifying the implementation of multi-slicing would be trickier.